### PR TITLE
Only load IPFS subresources within IPFS/IPNS schemes 

### PR DESCRIPTION
**Pull request by bbondy**
When a local node is installed, by visiting a page that allows you to load IPFS resources, it will then make you a host of any resources that the page loads.  To avoid this, this PR makes ipfs:// and ipns subresources only load when the top level main frame is also ipfs:// or ipns://.

This PR also fixes a problem where iframes would not load at all that were ipfs scheme. 

It also removes the external protocol handler that existed. 

Resolves https://github.com/brave/brave-browser/issues/12546
Resolves https://github.com/brave/brave-browser/issues/12931

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed).
- [ ] Requested a security/privacy review as needed.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: chromium_src/third_party/blink/common/loader/network_utils.cc [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "third_party/blink/public/common/loader/network_utils.h"
+#include "base/feature_list.h"
+#include "build/build_config.h"
+#include "net/net_buildflags.h"
+#include "services/network/public/cpp/is_potentially_trustworthy.h"
+#include "third_party/blink/public/common/features.h"
```
Comment (mkarolin) [inline chromium_src/third_party/blink/common/loader/network_utils.cc:11]: Nit: don't think these need to be included here.

File: chromium_src/third_party/blink/common/loader/network_utils.cc [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+#include "url/gurl.h"
+#include "url/url_constants.h"
```
Comment (mkarolin) [inline chromium_src/third_party/blink/common/loader/network_utils.cc:13]: Ditto.

File: chromium_src/third_party/blink/common/loader/network_utils.cc (hunk 1 above)
Comment (bbondy) [reply chromium_src/third_party/blink/common/loader/network_utils.cc]: Fixed in https://github.com/brave/brave-core/pull/7194/commits/afe446d035334d5272de6786c651e74a9edf1e5c

File: chromium_src/third_party/blink/common/loader/network_utils.cc (hunk 2 above)
Comment (bbondy) [reply chromium_src/third_party/blink/common/loader/network_utils.cc]: Fixed in https://github.com/brave/brave-core/pull/7194/commits/afe446d035334d5272de6786c651e74a9edf1e5c

**Review by mkarolin — APPROVED** [review]
`chromium_src` override LGTM

**Review by yrliou — APPROVED** [review]
LGTM!

File: chromium_src/third_party/blink/common/loader/network_utils.cc [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+#define IsURLHandledByNetworkService IsURLHandledByNetworkService_ChromiumImpl
+#include "../../../../../../third_party/blink/common/loader/network_utils.cc"
+#undef IsURLHandledByNetworkService
+
+namespace blink {
+namespace network_utils {
+
+bool IsURLHandledByNetworkService(const GURL& url) {
+  if (url.SchemeIs("ipns") || url.SchemeIs("ipfs")) {
```
Comment (iefremov) [inline chromium_src/third_party/blink/common/loader/network_utils.cc:16]: minor, but I'd try to transform all of "ipfs"/"ipns" to some constant defined somewhere at ipfs_constants.h

File: components/ipfs/ipfs_service_browsertest.cc [hunk 4]
```diff
@@ -266,60 +312,163 @@ IN_PROC_BROWSER_TEST_F(IpfsServiceBrowserTest, GetAddressesConfigServerError) {
...
-  ui_test_utils::NavigateToURL(browser(), GetURL("/simple.html"));
-  bool as_expected = false;
+  SetIPFSDefaultGatewayForTest(GetURL("b.com", "/"));
+  ui_test_utils::NavigateToURL(browser(), GetURL("b.com", "/iframe.html"));
   content::WebContents* contents =
       browser()->tab_strip_model()->GetActiveWebContents();
-  ASSERT_TRUE(ExecuteScriptAndExtractBool(
+
+  std::string location;
+  ASSERT_TRUE(ExecuteScriptAndExtractString(
```
Comment (iefremov) [inline components/ipfs/ipfs_service_browsertest.cc:373]: `ExecuteScriptAndExtractString` is deprecated in favor of `EvalJs` (see `browser_test_utils.h`). Also it should be possible to set an onload handler (i.e. avoid using the timer)

File: components/ipfs/ipfs_service_browsertest.cc [hunk 5]
```diff
@@ -266,60 +312,163 @@ IN_PROC_BROWSER_TEST_F(IpfsServiceBrowserTest, GetAddressesConfigServerError) {
...
+      ChildFrameAt(contents->GetMainFrame(), 0),
+      "const timer = setInterval(function () {"
+      "  if (document.readyState == 'complete') {"
+      "    clearInterval(timer);"
+      "    window.domAutomationController.send(window.location.href);"
+      "  }"
+      "}, 100);",
+      &location));
+
+  ASSERT_EQ(location, "chrome-error://chromewebdata/");
```
Comment (iefremov) [inline components/ipfs/ipfs_service_browsertest.cc:383]: I think `EXPECT_EQ` is more widespreaded across the codebase

**Review by iefremov — APPROVED** [review]

File: chromium_src/third_party/blink/common/loader/network_utils.cc (hunk 3 above)
Comment (bbondy) [reply chromium_src/third_party/blink/common/loader/network_utils.cc:16]: I think I'd have to patch in a dependency in that case no?

File: components/ipfs/ipfs_service_browsertest.cc (hunk 5 above)
Comment (bbondy) [reply components/ipfs/ipfs_service_browsertest.cc:383]: I think they are different, expect eq keeps running the test but assert eq fails the test instantly

File: components/ipfs/ipfs_service_browsertest.cc (hunk 4 above)
Comment (bbondy) [reply components/ipfs/ipfs_service_browsertest.cc:373]: ok thanks

File: components/ipfs/ipfs_service_browsertest.cc (hunk 5 above)
Comment (iefremov) [reply components/ipfs/ipfs_service_browsertest.cc:383]: They are indeed different, and the default choice is typically `EXPECT_EQ`. Not something important, especially for the last assertion in a test, just style

File: chromium_src/third_party/blink/common/loader/network_utils.cc (hunk 3 above)
Comment (iefremov) [reply chromium_src/third_party/blink/common/loader/network_utils.cc:16]: right, np then

**Pull request merged**